### PR TITLE
Referential validation of WDL. Defined a importResolver method that t…

### PR DESCRIPTION
**Pull request by gbggrant**
…akes an WDL import statement and returns the payload (WDL).

### CODE REVIEW

**Comment by dshiga** [comment]
This is very cool. Would it make sense to have the URI be agora:// instead of methods://? I'm not sure what all the pros and cons are, but just thought agora:// would be more specific.

**Comment by gbggrant** [comment]
That might make sense - will need to agree on this with cromwell.  We chose 'methods://' when we met on this last week - I think the purpose being to differentiate between methods and (someday) parameters?

File: src/main/scala/org/broadinstitute/dsde/agora/server/business/AgoraBusiness.scala [hunk 1]
```diff
@@ -38,4 +38,17 @@ object AgoraBusiness {
...
+  /**
+   * Used by cromwell to retrieve the WDL for tasks or workflows that have been pushed into the methods repository
+   * @param importString A string encoding the namespace, name and version of the method to look up (looks like: "methods://broad.grep.3")
+   * @return The payload of the method, if found, otherwise an empty string.
+   */
+  // Looks like "methods://broad.grep.3"
+  def importResolver(importString: String) : String = {
+    val importStringPattern = """^methods\:\/\/(\S+)\.(\S+).(\d+)""".r
+    importString match {
+      case importStringPattern(one, two, three) => AgoraBusiness.findSingle(one, two, three.toInt).getOrElse(new AgoraEntity()).payload.getOrElse("")
```
Comment (jacarey) [inline src/main/scala/org/broadinstitute/dsde/agora/server/business/AgoraBusiness.scala]: I would name these sub matches what they actually are (namespace, name, snapshotId)

File: src/main/scala/org/broadinstitute/dsde/agora/server/webservice/methods/MethodsAddHandler.scala [hunk 2]
```diff
@@ -32,7 +32,16 @@ class MethodsAddHandler extends Actor {
   }
 
   private def validatePayload(agoraEntity: AgoraEntity): Unit = {
-    WdlBinding.getAst(agoraEntity.payload.get, agoraEntity.name.get)
+//    WdlBinding.getAst(agoraEntity.payload.get, agoraEntity.name.get)
+
+    // Build the importString for testing...
+    val importString = "methods://" + agoraEntity.namespace.get + "." + agoraEntity.name.get + "." + 1
+    val fah = AgoraBusiness.importResolver(importString)
+    //    System.out.println(fah)
```
Comment (jacarey) [inline src/main/scala/org/broadinstitute/dsde/agora/server/webservice/methods/MethodsAddHandler.scala]: FAH!

File: src/test/scala/org/broadinstitute/dsde/agora/server/dataaccess/mongo/MethodsDbTest.scala [hunk 3]
```diff
@@ -35,4 +35,9 @@ class MethodsDbTest extends FlatSpec with AgoraDbTest with AgoraTestData {
 
     assert(entity1.snapshotId.get == entity2.snapshotId.get - 1)
   }
+
+  "Agora" should "not find an entity if it doesn't exist" in {
+    val entity = agoraDao.findSingle(testAgoraEntityNonExistent)
+    assert(entity == None)
```
Comment (jacarey) [inline src/test/scala/org/broadinstitute/dsde/agora/server/dataaccess/mongo/MethodsDbTest.scala]: entity.isEmpty

**Comment by jacarey** [comment]
:+1:

**Comment by gbggrant** [comment]
This pull request now contains all of the code to support referential validation, but since cromwell is not ready to support the import syntax in wdl, I have commented out the call in MethodsAddHandler.validatePayload that would do the referential validation (and also ignored one unit test).  With this, the code can be merged.

**Comment by jacarey** [comment]
:shipit:

**Pull request merged**